### PR TITLE
Protocol witness thunks don't need public linkage

### PR DESCRIPTION
**Pull request by slavapestov**
Follow-up to https://github.com/apple/swift/pull/8422. Giving the thunks public linkage is unnecessary, so we make them shared (if the witness table is going to be serialized) or private (otherwise).

This exposes a flaw in how we map AST-level accessibility to symbol visibility. We seem to allow protocol conformances that we really should reject. I added a workaround but we should consider making the relevant case an error in Swift 4 mode, since it's going to be an issue for ABI resilience and TBDGen.

### CODE REVIEW

File: test/SILGen/witness_accessibility.swift [hunk 1]
```diff
@@ -0,0 +1,42 @@
+// RUN: %target-swift-frontend -Xllvm -sil-full-demangle -emit-silgen %s | %FileCheck %s
+
+public protocol P {
```
Comment (slavapestov) [inline test/SILGen/witness_accessibility.swift]: @jrose-apple @huonw I invite you to marvel at the subtle complexity and delicious texture of this test case.

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@jrose-apple The thing I don't get is that removing 'public' from 'publicRequirement()' causes Sema to reject the conformance, but SILGen just ignores it anyway because R is 'fileprivate'.

I think is a consequence of how in SE-0025 (?) we decided that making things more public than the scope that contains them should be allowed, but ignored. I strongly feel this was a mistake.

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please benchmark

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please benchmark

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please test

**Comment by slavapestov** [comment]
@swift-ci Please benchmark

**Comment by slavapestov** [comment]
@swift-ci Please benchmark

**Comment by jrose-apple** [comment]
I think your analysis is correct. SE-0025 says this does not get a warning, and then something must be checking getFormalAccess where it should be looking at getFormalAccessScope. (The former really needs a rename.)

**Comment by jrose-apple** [comment]
Wow, those are some amazing wins and confusing losses.

**Comment by slavapestov** [comment]
I don't understand the results. There wasn't supposed to be a change.

**Comment by slavapestov** [comment]
@swift-ci Please benchmark

**Comment by jrose-apple** [comment]
Cost of interposable symbols, perhaps? Some things might be direct jumps instead of indirect now. But I don't know how things could get worse.

**Comment by slavapestov** [comment]
@jrose-apple Same result! I know @aschwaighofer is still iterating on performance of existentials, and *42* improvements seems like a decent outcome, so can we live with that one regression?

I'm glad I ran the benchmarks -- I was just expecting unexpected slowdowns or crashes.

**Pull request merged**